File: app/schemas/actas.py
# app/schemas/actas.py
"""
Schemas Pydantic v2 para el dominio actas administrativas.
"""

from datetime import date, datetime
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


class ActaGenerarRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    empleado_id: int
    incidencia_id: Optional[int] = None


class ActaFirmarRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    comentario: Optional[str] = None


FundamentoLegalActa = Literal[
    "Ley Federal del Trabajo",
    "Reglamento Interior de Trabajo",
]


class ActaCreateRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    empleado_id: int
    numero_empleado: str
    area_departamento: str
    supervisor_directo: str
    tipo_falta: str
    fundamento_legal: FundamentoLegalActa
    articulo_inciso: Optional[str] = None
    fecha_evento: date
    lugar_incidente: str
    descripcion_hechos: str
    personas_involucradas: Optional[str] = None
    testigos: Optional[str] = None
    responsable_rh: str
    evidencia: Optional[str] = None


class ActaAnularRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    motivo: Optional[str] = None


class ActaEditarRequest(BaseModel):
    model_config = {"str_strip_whitespace": True}

    tipo_falta: str
    fundamento_legal: FundamentoLegalActa
    articulo_inciso: Optional[str] = None
    fecha_evento: date
    lugar_incidente: str
    descripcion_hechos: str
    personas_involucradas: Optional[str] = None
    testigos: Optional[str] = None
    responsable_rh: str
    evidencia: Optional[str] = None


class ActaAprobacionResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    acta_id: int
    firmante_id: int
    rol_firmante: str
    firma_timestamp: Optional[datetime] = None
    comentario: Optional[str] = None


class ActaResponse(BaseModel):
    model_config = {"from_attributes": True}

    id: int
    empleado_id: int
    empleado_nombre: Optional[str] = None
    numero_empleado: Optional[str] = None
    puesto: Optional[str] = None
    area_departamento: Optional[str] = None
    supervisor_directo: Optional[str] = None
    tipo_falta: Optional[str] = None
    fundamento_legal: Optional[FundamentoLegalActa] = None
    articulo_inciso: Optional[str] = None
    fecha_evento: Optional[date] = None
    lugar_incidente: Optional[str] = None
    descripcion_hechos: Optional[str] = None
    personas_involucradas: Optional[str] = None
    testigos: Optional[str] = None
    responsable_rh: Optional[str] = None
    evidencia: Optional[str] = None
    incidencia_id: Optional[int] = None
    contenido_ia: Optional[str] = None
    ia_recomendacion: Optional[str] = None
    contenido_final: Optional[str] = None
    estado: str
    generado_por: int
    created_at: datetime
    aprobaciones: list[ActaAprobacionResponse] = []
    firmantes_pendientes: list[str] = []


class ActasPageResponse(BaseModel):
    """Listado paginado de actas (p. ej. Vista 360 por empleado)."""

    items: list[ActaResponse]
    total: int
    page: int
    page_size: int


class ActasDashboardMetricasResponse(BaseModel):
    """Conteos para tarjeta operativa RH (alineado a estados del listado de actas)."""

    en_proceso: int
    pendientes_firma: int


class ActaMejoraIaResponse(BaseModel):
    texto_mejorado: str


class ActaRagStatusResponse(BaseModel):
    status: str
    fresh: bool
    chroma_path: str
    collection: str
    chunks_indexed: int
    ollama: dict[str, Any]
    retrieval: dict[str, Any]
    expected_sources: list[str]
    manifest: dict[str, Any] | None = None
    errors: list[str] = Field(default_factory=list)

File: frontend/src/api/actas.ts
import { fetchWithAuth } from "./http.ts";

export type ActaCreatePayload = {
  empleado_id: number;
  numero_empleado: string;
  area_departamento: string;
  supervisor_directo: string;
  tipo_falta: string;
  fundamento_legal: "Ley Federal del Trabajo" | "Reglamento Interior de Trabajo";
  articulo_inciso?: string | null;
  fecha_evento: string;
  lugar_incidente: string;
  descripcion_hechos: string;
  personas_involucradas?: string | null;
  testigos?: string | null;
  responsable_rh: string;
  evidencia?: string | null;
};

export type ActaCreateResponse = {
  id: number;
  empleado_id: number;
  area_departamento: string | null;
  supervisor_directo: string | null;
  tipo_falta: string | null;
  fecha_evento: string | null;
};

export type ActaListItem = {
  id: number;
  empleado_id: number;
  empleado_nombre: string | null;
  numero_empleado: string | null;
  area_departamento: string | null;
  supervisor_directo: string | null;
  tipo_falta: string | null;
  fecha_evento: string | null;
  estado: "draft" | "pending_sign" | "signed" | "archived" | "cancelled";
  created_at: string;
};

export type ActaDetailResponse = {
  id: number;
  empleado_id: number;
  empleado_nombre: string | null;
  numero_empleado: string | null;
  puesto: string | null;
  area_departamento: string | null;
  supervisor_directo: string | null;
  tipo_falta: string | null;
  fundamento_legal: "Ley Federal del Trabajo" | "Reglamento Interior de Trabajo" | null;
  articulo_inciso: string | null;
  fecha_evento: string | null;
  lugar_incidente: string | null;
  descripcion_hechos: string | null;
  personas_involucradas: string | null;
  testigos: string | null;
  responsable_rh: string | null;
  evidencia: string | null;
  ia_recomendacion: string | null;
  estado: "draft" | "pending_sign" | "signed" | "archived" | "cancelled";
  created_at: string;
};

export type ActaUpdatePayload = {
  tipo_falta: string;
  fundamento_legal: "Ley Federal del Trabajo" | "Reglamento Interior de Trabajo";
  articulo_inciso?: string | null;
  fecha_evento: string;
  lugar_incidente: string;
  descripcion_hechos: string;
  personas_involucradas?: string | null;
  testigos?: string | null;
  responsable_rh: string;
  evidencia?: string | null;
};

export type ActaPageResponse = {
  items: ActaListItem[];
  next_cursor: number | null;
  total: number;
};

export type ActasDashboardMetricasResponse = {
  en_proceso: number;
  pendientes_firma: number;
};

export type ActaImproveWithIaResponse = {
  texto_mejorado: string;
};

export type ActaRagStatusResponse = {
  status: "ok" | "attention" | string;
  fresh: boolean;
  chroma_path: string;
  collection: string;
  chunks_indexed: number;
  ollama: {
    url: string;
    available: boolean;
    embedding_model: string;
    embedding_available: boolean;
  };
  retrieval: {
    top_k: number;
    score_threshold: number;
    balanced_documents: string[];
  };
  expected_sources: string[];
  manifest: {
    ingested_at?: string;
    total_chunks?: number;
    sources?: Array<{
      source: string;
      document_name: string;
      document_type: string;
      chunks: number;
      sha256: string;
    }>;
  } | null;
  errors: string[];
};

export type ActasFetchError = {
  status: number;
  detail: string;
};

async function readErrorDetail(res: Response): Promise<string> {
  const raw = await res.text();
  try {
    const j = JSON.parse(raw) as { detail?: unknown };
    if (typeof j.detail === "string") return j.detail;
  } catch {
    /* noop */
  }
  return raw || res.statusText || "Error";
}

function throwIfNotOk(res: Response, detail: string): never {
  const err: ActasFetchError = { status: res.status, detail };
  throw err;
}

export async function createActaAdministrativa(payload: ActaCreatePayload): Promise<ActaCreateResponse> {
  const res = await fetchWithAuth("/api/v1/actas", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaCreateResponse;
}

export async function getActasDashboardMetricas(): Promise<ActasDashboardMetricasResponse> {
  const res = await fetchWithAuth("/api/v1/actas/metricas-dashboard");
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActasDashboardMetricasResponse;
}

export async function getActasPage(params: {
  cursor?: number | null;
  limit?: number;
} = {}): Promise<ActaPageResponse> {
  const sp = new URLSearchParams();
  if (params.cursor != null) sp.set("cursor", String(params.cursor));
  sp.set("limit", String(params.limit ?? 100));
  const res = await fetchWithAuth(`/api/v1/actas?${sp.toString()}`);
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaPageResponse;
}

export async function getActaById(
  id: number,
  signal?: AbortSignal,
): Promise<ActaDetailResponse> {
  const res = await fetchWithAuth(`/api/v1/actas/${id}`, { signal });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaDetailResponse;
}

export async function improveActaWithIa(
  id: number,
  signal?: AbortSignal,
): Promise<ActaImproveWithIaResponse> {
  const res = await fetchWithAuth(`/api/v1/actas/${id}/mejorar-ia`, {
    method: "POST",
    signal,
  });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaImproveWithIaResponse;
}

export async function getActaRagStatus(signal?: AbortSignal): Promise<ActaRagStatusResponse> {
  const res = await fetchWithAuth("/api/v1/actas/rag/status", { signal });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaRagStatusResponse;
}

export async function reindexActaRag(signal?: AbortSignal): Promise<ActaRagStatusResponse> {
  const res = await fetchWithAuth("/api/v1/actas/rag/reindex", {
    method: "POST",
    signal,
  });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaRagStatusResponse;
}

export async function updateActaAdministrativa(
  id: number,
  payload: ActaUpdatePayload,
  signal?: AbortSignal,
): Promise<ActaDetailResponse> {
  const res = await fetchWithAuth(`/api/v1/actas/${id}/editar`, {
    method: "PUT",
    signal,
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaDetailResponse;
}

export async function approveActaAdministrativa(
  id: number,
  signal?: AbortSignal,
): Promise<ActaDetailResponse> {
  const res = await fetchWithAuth(`/api/v1/actas/${id}/aprobar`, {
    method: "PUT",
    signal,
  });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaDetailResponse;
}

export type ActaAnularPayload = {
  motivo?: string | null;
};

export async function anularActaAdministrativa(
  id: number,
  payload: ActaAnularPayload = {},
  signal?: AbortSignal,
): Promise<ActaDetailResponse> {
  const res = await fetchWithAuth(`/api/v1/actas/${id}/anular`, {
    method: "PUT",
    signal,
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ motivo: payload.motivo?.trim() || null }),
  });
  if (!res.ok) throwIfNotOk(res, await readErrorDetail(res));
  return (await res.json()) as ActaDetailResponse;
}
